platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ScopedFingerprintWriter.kt

```diff
@@ -21,38 +21,82 @@ import org.gradle.internal.serialize.graph.CloseableWriteContext
 import org.gradle.internal.serialize.graph.runWriteOperation
 import org.gradle.internal.serialize.graph.withPropertyTrace
 import java.io.Closeable
+import java.util.ArrayDeque
+import java.util.Queue
 
 
-internal
-class ScopedFingerprintWriter<T>(
+internal class ScopedFingerprintWriter<T>(
     private val writeContext: CloseableWriteContext
 ) : Closeable {
+    private class PendingWrite<out T>(val value: T, val trace: PropertyTrace?)
+
+    private val pendingWrites: Queue<PendingWrite<T>> = ArrayDeque()
+
+    private var isWriting = false
+
     override fun close() {
         // we synchronize access to all resources used by callbacks
         // in case there was still an event being dispatched at closing time.
         synchronized(writeContext) {
-            unsafeWrite(null)
+            require(!isWriting && pendingWrites.isEmpty()) {
+                "Incomplete write detected"
+            }
+
+            writeValueLocked(null)
```

Comment:
🤔 This `Locked` suffix reads backward to me. `writeValueLocked` makes me think the method will acquire the lock to write value.

Reply:
> 🤔 This `Locked` suffix reads backward to me. `writeValueLocked` makes me think the method will acquire the lock to write value.

I've picked this convention when working with AOSP, but I see how it can be confusing.

Reply:
I'd still use some suffix or prefix to differentiate this method from the synchronizing one but we can come up with something later.